### PR TITLE
Use cli parser with auto-generated help

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by matklad** [comment]
Time to compile xtask:

```
// before
real 9.65s
user 8.89s
sys  0.82s
rss  295476k

// after
real 10.19s
user 9.42s
sys  0.83s
rss  296840k
```

File: xtask/src/flags.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        cmd dist {
+            optional --nightly
+            optional --client version: String
+        }
+        cmd metrics {
+            optional --dry-run
+        }
+        /// Builds a benchmark version of rust-analyzer and puts it into `./target`.
+        cmd bb
+            required suffix : String
```
Comment (lnicola) [inline xtask/src/flags.rs:49]:
```suggestion
            required suffix: String
```

**Comment by matklad** [comment]
binary size diff with `-z`, strip, full lto, and single codgen unit:

```
λ exa -l target/
.rw-r--r--  177 matklad  1 Mar 22:25 CACHEDIR.TAG
drwxr-xr-x    - matklad  1 Mar 22:25 debug
drwxr-xr-x    - matklad  1 Mar 22:28 release
.rwxr-xr-x 928k matklad  1 Mar 22:28 xtask-pico
.rwxr-xr-x 932k matklad  1 Mar 22:28 xtask-xflags
```

**Comment by matklad** [comment]
bors r+

I am not feeling completely easy with switching to my own arg-parsing library, but what one can do if alternatives are not that compelling? :)

**Comment by matklad** [comment]
cc @razrfalcon, you might, or might not like the idea behind [xflags](https://github.com/matklad/xflags). It is significantly more bloated than pico-args, but still significantly less bloated than clap or gumdrop. Of course, while the idea is interesting, the implementation is not production ready yet.

**Comment by RazrFalcon** [comment]
@matklad Yeah, proc-macros is no go for me. It hurts compilation times too much. Yes, they've become way better recently, but it still at least 5x slower.

PS: I wrote pico-args just for myself. I'm honestly surprised that anyone actually using it.

**Comment by lnicola** [comment]
They aren't that slow if you skip syn. See the benchmark above.

**Comment by RazrFalcon** [comment]
I didn't know you can do this. Anyway, I don't really like the api. I would prefer struct attributes, just like structops/gumdrop/argh.

**Comment by matklad** [comment]
Yup, surprisingly, proc macros themselves are not that bad for compile time. I’ve learned about that couple of months ago, here’s a small benchmark setup if you are curious: https://github.com/matklad/xshell/tree/master/cbench

proc macros are usually slow for two reasons:

* they require syn, which is, understandably, slow to compile itself, and becomes sequential bottleneck
* they generate a ton of complex code (serde)

If you don’t do these two things, the cost becomes much lower (but is still not free, as compiling proc macro itself means at least one extra rustc invocation).

But yeah, that is one of the reasons why xflags is not derive: with derive, you need to parse rust syntax (instead of simplified macro-specific syntax), and that’s awkward to do without syn. The second reason is that, with sub commands, it’s not really convenient to describe the cli as a *single* struct.

**Comment by RazrFalcon** [comment]
I knew that proc-macros is basically a shared library, but I thought that syn is mandatory. And it takes the largest time chunk.
Can't you parse structs manually? Maybe using some sort of pico-syn (wink).

Also, maybe it's just me, but autogenerated help is always awful. It just not flexible enough. But less error-prone, obviously.

PS: doesn't this kind of macro breaks autocomplete?

**Comment by matklad** [comment]
>Can't you parse structs manually? 

You certainly can, it's just a bit fiddly. Copy-pasting nanoserde is probably the way to go: https://github.com/not-fl3/nanoserde/blob/master/derive/src/parse.rs

> Also, maybe it's just me, but autogenerated help is always awful. It just not flexible enough. But less error-prone, obviously.

Yeah, that's absolutely true! But for me personally, most of the CLI tools I write are for "internal use", so I just don't get to writing proper helps. Auto-generated help is a lot better than no help at all, and doesn't take any effort. 

>PS: doesn't this kind of macro breaks autocomplete?

rust-analyzer can handle this, if proc macro support is enabled (we are just in the process of enabling it by default). IntelliJ Rust, [as of 3 days ago](https://github.com/intellij-rust/intellij-rust/pull/6564) should be able to handle that to.

But this is one of the motivations why the API is semi-proc macro, and [generates some code in place](https://github.com/rust-analyzer/rust-analyzer/pull/7835/files#diff-24fc1aac9c36abb39a768ec4e42c31cf85d2f9de972313c3528b193134d32dfbR54-R139). That way, even IDE without proc macro support should give correct completions. And for the user themselves it's also useful to see what the heck is actually generated.

**Comment by RazrFalcon** [comment]
Thanks!

File: xtask/src/main.rs [hunk 2]
```diff
@@ -32,145 +32,63 @@ use crate::{
...
 
-            let client_opt = args.opt_value_from_str("--client")?;
-
-            finish_args(args)?;
+            let client_bin = flags.code_bin.map(|it| it.parse()).transpose()?;
 
             InstallCmd {
-                client: if server { None } else { Some(client_opt.unwrap_or_default()) },
-                server: if client_code { None } else { Some(ServerOpt { malloc }) },
+                client: if flags.server { None } else { client_bin },
```
Comment (vincentisambart) [inline xtask/src/main.rs:66]: With this change, if you do not specify `code_bin` explicitly, the client will not be installed. Was that on purpose?

**Pull request merged**